mirgecom/flux.py

```diff
@@ -0,0 +1,77 @@
+""":mod:`mirgecom.flux` provides inter-facial flux routines.
+
+Numerical Flux Routines
+^^^^^^^^^^^^^^^^^^^^^^^
+
+.. autofunction:: lfr_flux
+"""
+
+__copyright__ = """
+Copyright (C) 2021 University of Illinois Board of Trustees
+"""
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+
+def lfr_flux(q_tpair, compute_flux, normal, lam):
+    r"""Compute Lax-Friedrichs/Rusanov flux after [Hesthaven_2008]_, Section 6.6.
+
+    The Lax-Friedrichs/Rusanov flux is calculated as:
+
+    .. math::
+
+        f_{\mathtt{LFR}} = \frac{1}{2}(\mathbf{f}^{+} + \mathbf{f}^{-}) \cdot
+        \hat{n} + \frac{\lambda}{2}(q^{-} - q^{+}),
+
+    where $f^-, f^+$, and $q^-, q^+$ are the fluxes and scalar solution components on
+    the interior and the exterior of the face on which the LFR flux is to be
+    calculated. The face normal is $\hat{n}$, and $\lambda$ is the user-supplied
+    jump term coefficient.
+
+    Parameters
+    ----------
+    compute_flux:
+
+        function should return ambient dim-vector fluxes given *q* values
+
+    q_tpair: :class:`grudge.sym.TracePair`
+
+        Trace pair for the face upon which flux calculation is to be performed
+
+    normal: numpy.ndarray
+
+        object array of :class:`meshmode.dof_array.DOFArray` with outward-pointing
+        normals
+
+    lam: :class:`meshmode.dof_array.DOFArray`
+
+        lambda parameter for Lax-Friedrichs/Rusanov flux
+
+    Returns
+    -------
+    numpy.ndarray
+
+        object array of :class:`meshmode.dof_array.DOFArray` with the
+        Lax-Friedrichs/Rusanov flux.
+    """
+    flux_avg = 0.5*(compute_flux(q_tpair.int)
+                    + compute_flux(q_tpair.ext))
+    return flux_avg @ normal - 0.5*lam*(q_tpair.ext - q_tpair.int)
```

Comment:
How about:
```suggestion
def lfr_flux(q_tpair, flux_func, normal, lam):
    r"""Compute Lax-Friedrichs/Rusanov flux after [Hesthaven_2008]_, Section 6.6.

    The Lax-Friedrichs/Rusanov flux is calculated as:

    .. math::

        f_{\mathtt{LFR}} = \frac{1}{2}(\mathbf{F}(q^-) + \mathbf{F}(q^+)) \cdot
        \hat{n} + \frac{\lambda}{2}(q^{-} - q^{+}),

    where $q^-, q^+$ are the scalar solution components on the interior and the
    exterior of the face on which the LFR flux is to be calculated, $\mathbf{F}$ is
    the vector flux function, $\hat{n}$ is the face normal, and $\lambda$ is the
    user-supplied jump term coefficient.

    Parameters
    ----------
    flux_func:

        function should return ambient dim-vector fluxes given *q* values

    q_tpair: :class:`grudge.sym.TracePair`

        Trace pair for the face upon which flux calculation is to be performed

    normal: numpy.ndarray

        object array of :class:`meshmode.dof_array.DOFArray` with outward-pointing
        normals

    lam: :class:`meshmode.dof_array.DOFArray`

        lambda parameter for Lax-Friedrichs/Rusanov flux

    Returns
    -------
    numpy.ndarray

        object array of :class:`meshmode.dof_array.DOFArray` with the
        Lax-Friedrichs/Rusanov flux.
    """
    flux_avg = 0.5*(flux_func(q_tpair.int)
                    + flux_func(q_tpair.ext))
    return flux_avg @ normal - 0.5*lam*(q_tpair.ext - q_tpair.int)
```

(Mostly just changing the notation in the documentation, and renaming `compute_flux` to `flux_func` to make it marginally less jarring.)

Reply:
(Easier to read version: [link](https://github.com/majosm/mirgecom/commit/cecd3d9e72e71c66146265cab6d6bdcb36bf310c))